src/cryptonote_basic/cryptonote_format_utils.cpp

```diff
@@ -1102,4 +1141,63 @@ namespace cryptonote
     return key;
   }
 
+  //---------------------------------------------------------------
+  #define CHACHA8_KEY_TAIL 0x8c
+  bool generate_chacha_key_from_secret_keys(const account_keys &keys, crypto::chacha_key &key)
+  {
+    const crypto::secret_key &view_key = keys.m_view_secret_key;
+    const crypto::secret_key &spend_key = keys.m_spend_secret_key;
+    tools::scrubbed_arr<char, sizeof(view_key) + sizeof(spend_key) + 1> data;
+    memcpy(data.data(), &view_key, sizeof(view_key));
+    memcpy(data.data() + sizeof(view_key), &spend_key, sizeof(spend_key));
+    data[sizeof(data) - 1] = CHACHA8_KEY_TAIL;
+    crypto::generate_chacha_key(data.data(), sizeof(data), key);
+    return true;
+  }
+
+  //---------------------------------------------------------------
+  crypto::public_key get_subaddress_spend_public_key(const cryptonote::account_keys& keys, const cryptonote::subaddress_index& index)
+  {
```

Comment:
Aren't thee functions copied from somewhere else? Can the other copies be removed?

Reply:
Original code from generate_chacha_key from wallet2.

moved here to be able work on provided account_keys.  generate_chacha_key now call the device function witch call   generate_chacha_key_from_secret_keys

Same transformation for others respecting the  namespace of similar subaddress related functions.